xdsl/interpreters/experimental/wgsl_printer.py

```diff
@@ -0,0 +1,248 @@
+                    auth = "read_write"
+            if isinstance(arg.type, builtin.Float32Type):
+                arg_type = "f32"
+            elif isinstance(arg.type, builtin.IndexType):
+                arg_type = "u32"
+            elif isinstance(arg.type, MemRefType):
+                memref_typ = cast(MemRefType[Attribute], arg.type)
+                arg_type = f"array<{memref_typ.element_type}>"
+            arguments = f"""
+    @group(0) @binding({arg.index})
+    var<storage,{auth}> {self.wgsl_name(arg)}: {arg_type};
+"""
+            out_stream.write(arguments)
+
+        out_stream.write(
+            """
+    @compute
+    @workgroup_size(1)
+    fn main(@builtin(global_invocation_id) global_invocation_id : vec3<u32>,
+    @builtin(workgroup_id) workgroup_id : vec3<u32>,
+    @builtin(local_invocation_id) local_invocation_id : vec3<u32>,
+    @builtin(num_workgroups) num_workgroups : vec3<u32>) {
+"""
+        )
+        for operation in op.body.ops:
+            self.print(operation, out_stream)
+        out_stream.write(
+            """
+            }
+            """
+        )
+
+    @print.register
+    def _(self, op: gpu.ReturnOp, out_stream: IO[str]):
+        pass
+
+    @print.register
+    def _(self, op: gpu.BlockIdOp, out_stream: IO[str]):
+        dim = str(op.dimension.value.param).replace('"', "")
+        name_hint = self.wgsl_name(op.result)
+        out_stream.write(
+            f"""
+        let {name_hint}: u32 = workgroup_id.{dim};"""
+        )
+
+    @print.register
+    def _(self, op: gpu.ThreadIdOp, out_stream: IO[str]):
+        dim = str(op.dimension.value.param).replace('"', "")
+        name_hint = self.wgsl_name(op.result)
+        out_stream.write(
+            f"""
+        let {name_hint}: u32 = local_invocation_id.{dim};"""
+        )
+
+    @print.register
+    def _(self, op: gpu.GridDimOp, out_stream: IO[str]):
+        dim = str(op.dimension.value.param).replace('"', "")
+        name_hint = self.wgsl_name(op.result)
+        out_stream.write(
+            f"""
+        let {name_hint}: u32 = num_workgroups.{dim};"""
+        )
+
+    @print.register
+    def _(self, op: gpu.BlockDimOp, out_stream: IO[str]):
+        dim = str(op.dimension.value.param).replace('"', "")
+        name_hint = self.wgsl_name(op.result)
+        out_stream.write(
+            f"""
+        let {name_hint}: u32 = local_invocation_id.{dim};"""
+        )
+
+    @print.register
+    def _(self, op: gpu.GlobalIdOp, out_stream: IO[str]):
+        dim = str(op.dimension.value.param).replace('"', "")
+        name_hint = self.wgsl_name(op.result)
+        out_stream.write(
+            f"""
+        let {name_hint}: u32 = global_invocation_id.{dim};"""
+        )
+
+    @print.register
+    def _(self, op: memref.Load, out_stream: IO[str]):
+        memref_type = cast(MemRefType[Attribute], op.memref.type)
+        memref_dimension = memref_type.get_num_dims()
+        memref_size = memref_type.get_shape()
```

Comment:
You should assert that it has a fixed size, otherwise you may get some funny surprises where `memref<100x?>` has size -100 :sweat_smile: